### ISSUE TITLE
Typescript type declaration issue

### ISSUE THREAD

**Issue by blackshadev** [express-promise-router/express-promise-router#50]
Hi I have an issue with using your supplied type definition.

Typescript wants me to use your exported 'default' function instead of using your whole module as a function.

I got the following error
```
error TS2349: Cannot invoke an expression whose type lacks a call signature. Type 'typeof 'express-promise-router'' has no compatible call signatures.
```
On the following lines
```
import * as Router from "express-promise-router";
const router = Router() as express.Router;
```
Looking at your index.d.ts I suggest you take a look at [this example ](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-function-d-ts.html) by the TS team, which exports a specific function declaration as the whole module.

**Comment by mormahr** [comment]
Hi

coming from FlowType and babel my approach would be to `import Router from "express-promise-router"` but now that i've tested it i realize that TypeScript handles this import differently from babel. I'll look into this, but i'd rather export a default property, because `import * as Router` doesn't seem to be semantically right.

My ideal working example would be this:

```ts
import * as express from "express"
import Router from "express-promise-router"
const router = Router()

router.get("/", function (req, res) {
    res.send("Hi!")
})

const app = express()
app.use(router)
app.listen(3000)
```

This would require the following source modification:
```diff
Index: lib/express-promise-router.js
<+>UTF-8
===================================================================
--- lib/express-promise-router.js	(date 1516408799000)
+++ lib/express-promise-router.js	(date 1516745458000)
@@ -110,4 +110,6 @@
     return me;
 };
 
+PromiseRouter.default = PromiseRouter;
+
 module.exports = PromiseRouter;
```
I'm pretty sure this won't cause any harm for other users and use-cases but i'm cautious to not break anything anyway.

Please tell me what you think about it. And @bangbang93 @jcristovao feel free to chime in as well!

Thanks
Moritz

**Comment by bangbang93** [comment]
dts file also need some modification
```diff
diff --git a/index.d.ts b/index.d.ts
index 8d964a1..2e5806a 100644
--- a/index.d.ts
+++ b/index.d.ts
@@ -4,4 +4,6 @@ declare module 'express-promise-router' {
     function PromiseRouter(options?: RouterOptions): Router
 
     export default PromiseRouter
+
+    export = PromiseRouter
 }
```
to support `import * as Router from 'express-promise-router`

**Comment by mormahr** [comment]
Yes but like i said: I don‘t think `import * as Router` is the right way to do it. And `import Router` works with the current d.ts as far as i can see.

The `* as` syntax is to collect named exports in a dictionary object.

```js
export const foo = 5
export const bar = 10
```

```js
import * as obj from “./a“

obj.foo
```

but we don‘t export multiple named things.

Please correct me if i‘m wrong and please tell me how you feel about adding the default property to our export / to the Router function.

**Comment by bangbang93** [comment]
I agree with you that using `import Router from 'express-promise-router'` is much better than `import * as Router`

In fact I'm using `const router = require('express-promise-router')();` in typescript because is much shorter.
But after replacing with
```javascript
import Router from 'express-promise-router'
const router = Router();
```
I got an error
> const router = Router();
                     ^
TypeError: express_promise_router_1.default is not a function

So actually we need a default property to export Router function in js code

**Comment by mormahr** [comment]
Yes, i'd implement it the way i said in my [first comment](https://github.com/express-promise-router/express-promise-router/issues/50#issuecomment-359950205). But to be sure to not break anything i'll add tests that use native CommonJS requires, babel transpiled imports and TypeScript transpiled imports. Are there any other transpilers that we need to test?

**Comment by mormahr** [comment]
I started working on this in #51 but there is still some work to do. Best case would be finishing this tomorrow. But i don't know how much time i have and need exactly.

**Comment by blackshadev** [comment]
Hi,

I actually didn't got any emails about your reactions, so sorry for my late reply.

As of ts 2.7 they added the `import Router from "express-promise-router";` syntax . Which DOES depends on the export as default behavior.   Please consult the [what is new page](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html) of ts for more info.
So maybe you want to leave the default export...

**Comment by SimonSchick** [comment]
Same issue, the `new` signature is broken now.

My declaration was as following:

```
declare module 'express-promise-router' {
  import { Router } from 'express';
  interface ExpressPromiseRouter extends Router {}
  class ExpressPromiseRouter implements Router {}
  namespace ExpressPromiseRouter {}
  export = ExpressPromiseRouter;
}
```

**Comment by sanyatuning** [comment]
I created a pull request to fix this issue #53

**Comment by mormahr** [comment]
Hi!

I'm sorry i'm ill at the moment and can't work on this for a few more days. 

@sanyatuning this is not what we are looking for semantically. Please see #51. We do want a default export and #51 makes that happen. There are just two missing tests (for babel-commonjs and for regular commonjs) that i haven't had the time to write jet.

@SimonSchick what do you mean with `new` signature? `var router = new PromiseRouter()`? Because that shouldn't be used anyway.

**Comment by Justus-Maier** [comment]
Just hit this issue. Please accept #53, it fixes the issue for me!

**Comment by Justus-Maier** [comment]
> var router = new PromiseRouter()? Because that shouldn't be used anyway.
If the default export is called PromiseRouter (capital P), in common JS coding standards this means it's a constructor and should be called with "new".

**Comment by SimonSchick** [comment]
@mormahr as @Justus-Maier suggested it's implied you should call it with new, it also _works_ with new and removing the ability to call it that way should be considered a breaking change.

**Comment by Justus-Maier** [comment]
In meantime, could you please provide an example how the import/initialisation currently has to look like to avoid TS2349? Or please release a fixed version.

**Comment by blackshadev** [comment]
I created a file which does the import and typing for me. See example below.

Do mind that I didn't update the package for a time , but this was my work-around since the creation of this issue.

```
import * as pRouter from "express-promise-router";
import { Router as eRouter } from "express";
export const Router = (pRouter as any) as () => eRouter;
```

**Comment by sant123** [comment]
Any advance on this? Just asking 😋

**Comment by Justus-Maier** [comment]
I'll go with blackshadevs workaround now, as I'm getting tired of fixing this manually every time I use a yarn command. 
@mormahr why can't you just publish a simple fix and then worry about default imports in a major version later?

**Comment by sant123** [comment]
I'll be happy to use this library but meanwhile I'm using the old school way... An async function with a try catch clause. If an error occurs I just pass it through the next function.

**Comment by chanlito** [comment]
@sant123 that's a terrible DX.

**Comment by sant123** [comment]
Sorry 🤷‍♂️

**Comment by mormahr** [comment]
@Justus-Maier 
> why can't you just publish a simple fix and then worry about default imports in a major version later?

This is why things break. It's always "just accept it, it'll be fine". It never is. This whole TypeScript thing has been pretty frustrating. I'm not a TypeScript guy so i trusted contributors to thoroughly test, but things broke.

@SimonSchick There is no functionality being removed, only added. Namely that the commonjs export now has an additional property `default` so that babel and ts imports work as intended. To ensure this i spent some time adding tests in #51 which i would like to merge today. Can we discuss adding a `new` signature in a new issue please? Using new technically has always worked (because it always does in js unless you specifically prohibit it), but was never the way the readme "defined" the usage. I'm actually not against `new`, it's just that when i took over maintenance it was like this and i haven't changed it since. But i think this is not the core of this issue, so let's start a new one. ☺️

@sanyatuning I advocate proposing PRs but please make sure unit tests are passing. We can discuss your change on the PR if you'd like, but it alters the specified behavior iirc.

To end this: **Please go ahead and test #51 (available as v3.0.2-rc.1)** and tell me if this fixes your various problems.
It should work with these snippets:

TypeScript **and** Babel (`transform-es2015-modules-commonjs`)
```typescript
import Router from 'express-promise-router';
const router = Router();
```

basic js:
```js
var Router = require('express-promise-router.js');
var router = Router();
```

So can we be happy again now? 😜

**Comment by pluma** [comment]
FWIW if you don't like `import * as Router from "..."` (which is weird) you can also use `import Router = require("...")` (which is invalid JS but valid TS and mirrors the `export = ...` syntax).

The only problem of course is that this doesn't work with the provided typings.

**Comment by mormahr** [comment]
@pluma It's not weird. It's wrong. There is no `import * as Router` in the example. It's `import Router` which imports the default export.

I don't know why so many people are writing `import * as Router`. That roughly translates to `grab all the named exports and bundle them as an object`, but we export a single function as the default export. That `import * as` even works is only because of compatibility reasons with the commonjs approach to export an object that contains the "named exports". 

@pluma Is `import Router from "express-promise-router"` succinct enough for you? If not, please feel free to open a new issue regarding the support of a new syntax.

**Comment by pluma** [comment]
@mormahr there's no reason to assume things. Open up this: https://www.typescriptlang.org/play/index.html

And enter the following:

```typescript
import Router from 'express-promise-router';

const router = Router();

export default router;
```

Notice how it generates this line:

```js
var router = express_promise_router_1.default();
```

This will result in a runtime exception unless the module loader does something special to add a `default` attribute to the imported module. `tsc` doesn't do that, so the result is either a compile time type error or a runtime error.

**Comment by pluma** [comment]
BTW, here's what I'm seeing `tsc` generate for Node:

```js
const express_promise_router_1 = require("express-promise-router");
const router = express_promise_router_1.default();
```

**Comment by mormahr** [comment]
Please try v3.0.2-rc.1 which has the fix #51

**Comment by pluma** [comment]
@mormahr yupp, that works fine. Anything stopping you from going ahead with that release?

**Comment by mormahr** [comment]
Yes and no, i just wanted people to test it, but that didn't seem to work too well 😃
I'll release it in 20min.

**Comment by mormahr** [comment]
Yeah don't trust me on timelines. It's published as v3.0.2. I published as a bug fix even though it's technically a feature because it should have been like that since v3.0.0.

**Comment by Justus-Maier** [comment]
"new" keyword doesn't, but this works fine - thanks a lot!
```
import createPromiseRouter from 'express-promise-router';
const router = createPromiseRouter();
```

**Comment by mmouterde** [comment]
The fix does not include the index.d.ts change from https://github.com/express-promise-router/express-promise-router/pull/54

Please reopen this issue.

**Comment by 52cik** [comment]
Same issue.

My declaration was as following:

```ts
declare module 'express-promise-router' {
    import { Router, RouterOptions } from 'express';

    // function PromiseRouter(options?: RouterOptions): Router;

    // export default PromiseRouter;

    interface IPromiseRouter {
        (options?: RouterOptions): Router;
        default: IPromiseRouter;
    }

    const PromiseRouter: IPromiseRouter;
    export = PromiseRouter;
}
```

```js
// Support for default import
import Router from 'express-promise-router';
// support for cjs import
const Router = require('express-promise-router');
```

**Comment by douggr** [comment]
This is fixed in `3.0.3`.

```ts
import * as express from "express";
import PromiseRouter from "express-promise-router";
import { Request, Response } from "express";

export const router = PromiseRouter();


router.get(
  "/",
  async(req: Request, res: Response): Promise<string> => {
    return Promise.resolve("next");
  },
  async(req: Request, res: Response): Promise<Response> => {
    return res.jsonp("yay!");
  },
);


export const server = express().use(router);
```
